File: viewer/src/types/catalog.ts
/**
 * TypeScript types for the Peblo TV published catalogue.
 *
 * These mirror the Pydantic schemas in
 * `backend/app/schemas/catalog.py` exactly. The Viewer is a
 * read-only consumer of the published snapshot; it MUST NOT
 * invent additional shape, and it MUST NOT call any
 * admin / CMS endpoint to reconstruct this data.
 *
 * Source of truth (backend):
 *   - GET /catalog          -> Catalog
 *   - GET /catalog/search   -> SearchResponse
 *
 * The catalogue document is built by the backend publisher from
 * the live database, frozen onto object storage, and returned as a
 * single JSON blob. The Viewer therefore works with the entire
 * catalogue (so it can render Netflix-style section rows locally)
 * AND has a dedicated search endpoint that runs server-side
 * filtering (so search never pulls the full doc just to filter
 * client-side — except the home page, which needs everything to
 * render its section rows).
 */

export interface ArtworkItem {
  type: string;
  url: string;
  width?: number | null;
  height?: number | null;
  mime_type?: string | null;
}

export type ArtworkMap = Record<string, ArtworkItem>;

export interface CatalogEpisode {
  content_group: string;
  languages: string[];
  canonical_language: string;
  title: string;
  episode_number: number;
  duration?: number | null;
  artwork: ArtworkMap;
}

export interface CatalogTrailer extends CatalogEpisode {}

export interface CatalogSeason {
  season_number: number;
  episodes: CatalogEpisode[];
}

export interface CatalogShow {
  id: number;
  slug: string;
  title: string;
  synopsis?: string | null;
  section: string;
  categories: string[];
  trailers: CatalogTrailer[];
  seasons: CatalogSeason[];
}

export interface CatalogSection {
  key: string;
  shows: CatalogShow[];
}

export interface Catalog {
  version: number;
  published_at: string | null;
  publish_run_id: number | null;
  sections: CatalogSection[];
}

export interface SearchQuery {
  q?: string | null;
  category?: string | null;
  language?: string | null;
  section?: string | null;
}

export interface SearchResponse {
  query: SearchQuery;
  count: number;
  results: CatalogShow[];
}

/**
 * Well-known section keys. The backend publisher always emits
 * sections in this order. The Viewer renders them in the same
 * order and labels them consistently.
 */
export const SECTION_KEYS = ['featured', 'series', 'minisodes', 'songs'] as const;
export type SectionKey = (typeof SECTION_KEYS)[number];

/**
 * Human-friendly labels for the known section keys. The Viewer
 * only renders rows for these four; any other section key
 * returned by the API still renders with a derived label
 * (`titleCase(key)`).
 */
export const SECTION_LABELS: Record<string, string> = {
  featured: 'Featured',
  series: 'Series',
  minisodes: 'Minisodes',
  songs: 'Songs',
};

export function sectionLabel(key: string): string {
  if (SECTION_LABELS[key]) return SECTION_LABELS[key]!;
  const k = key.trim();
  if (!k) return 'More';
  return k.charAt(0).toUpperCase() + k.slice(1);
}

File: backend/app/schemas/catalog.py
"""
Pydantic schemas for the published catalogue (Phase 6).

Design notes:

* Frozen on object storage by POST /admin/catalog/publish.
* Sections: featured, series, minisodes, songs.
* content_group collapses into ONE episode with languages list.
* Season 0 -> show.trailers (never in show.seasons).
* artwork contains ONLY types that actually exist.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ArtworkItem(BaseModel):
    model_config = ConfigDict(extra="forbid")
    type: str
    url: str
    width: int | None = None
    height: int | None = None
    mime_type: str | None = None


class EpisodeArtwork(BaseModel):
    model_config = ConfigDict(extra="forbid")
    poster: ArtworkItem | None = None
    banner: ArtworkItem | None = None
    thumbnail: ArtworkItem | None = None

    def to_dict(self) -> dict[str, Any]:
        return self.model_dump(exclude_none=True)


class CatalogEpisode(BaseModel):
    model_config = ConfigDict(extra="forbid")
    content_group: str
    languages: list[str] = Field(min_length=1)
    canonical_language: str
    title: str
    episode_number: int = Field(ge=0)
    duration: int | None = Field(default=None, ge=0)
    artwork: dict[str, ArtworkItem] = Field(default_factory=dict)


class CatalogTrailer(CatalogEpisode):
    model_config = ConfigDict(extra="forbid")


class CatalogSeason(BaseModel):
    model_config = ConfigDict(extra="forbid")
    season_number: int = Field(ge=1)
    episodes: list[CatalogEpisode]


class CatalogShow(BaseModel):
    model_config = ConfigDict(extra="forbid")
    id: int = Field(ge=1)
    slug: str
    title: str
    synopsis: str | None = None
    section: str
    categories: list[str] = Field(default_factory=list)
    trailers: list[CatalogTrailer] = Field(default_factory=list)
    seasons: list[CatalogSeason] = Field(default_factory=list)


class CatalogSection(BaseModel):
    model_config = ConfigDict(extra="forbid")
    key: str
    shows: list[CatalogShow] = Field(default_factory=list)


class Catalog(BaseModel):
    model_config = ConfigDict(extra="forbid")
    version: int = 1
    published_at: datetime
    publish_run_id: int | None = None
    sections: list[CatalogSection] = Field(default_factory=list)


def empty_catalogue() -> dict[str, Any]:
    return {
        "version": 1,
        "published_at": None,
        "publish_run_id": None,
        "sections": [],
    }


__all__ = [
    "ArtworkItem", "Catalog", "CatalogEpisode",
    "CatalogSection", "CatalogSeason", "CatalogShow",
    "CatalogTrailer", "EpisodeArtwork", "empty_catalogue",
]
